### ISSUE TITLE
"Regex too complicated" error when logging HTTP response with long body

### ISSUE THREAD

**Issue by TrangOul** [jongpie/NebulaLogger#587]
### Package Edition of Nebula Logger

Unlocked Package

### Package Version of Nebula Logger

v4.12.0

### New Bug Summary

Stack trace:
```
15:03:27.940 (6272690856)|FATAL_ERROR|System.LimitException: Regex too complicated

Class.LogEntryEventBuilder.applyDataMaskRules: line 928, column 1
Class.LogEntryEventBuilder.setHttpResponseDetails: line 584, column 1
```

My code:
```apex
LogEntryEventBuilder logBuilder = Logger.fine(logTitle)
       .setHttpRequestDetails(this.request)
       .addTag('Callout');
this.response = new Http().send(this.request);

//shortening body to pinpoint exact length causing error
/*
String body = this.response.getBody();
final Integer sub = 457204;
body = body.substring(0, sub);
this.response.setBody(body);
*/

logBuilder.setHttpResponseDetails(this.response); //this line fails - long before Logger.saveLog()
(...)
Logger.saveLog();
```

A simpler snippet to replicate the issue:
```apex
HttpResponse resp = new HttpResponse();
resp.setBody('1'.repeat(500006));
Logger.fine('log').setHttpResponseDetails(resp);
```

This happens only when the HTTP response body size is large.
In my case, it was initially over 5 million characters long. By substringing and binary search (shown above) I was able to figure out that the error occurs when body is at least 457 204 chars long in the original code and 500 006 in the simpler snippet (the discrepancy is likely because of Unicode characters in the real response body).
Such a long body won't be saved in the log anyway because of the much lower limit of the long text field (131 072 chars), so maybe it should be truncated before masking (or other kind of processing)? Other logged values should be treated the same way.

Moreover, I do not have any masking rules besides the default ones (`MastercardCreditCardNumber`, `SocialSecurityNumber`, `VisaCreditCardNumber`).

**Comment by dvdjg** [comment]
I have a similar problem.

My current workaround is to limit the string size for any long string in the **LogEntryEventBuilder.applyDataMaskRules** function:

    private String applyDataMaskRules(String dataInput) {
        **if (String.isNotBlank(dataInput)) {
            // Trim the input data to the max size for a long text in Salesforce
            // This will also limit the replaceAll regex complexity
            dataInput = dataInput.left(131072);
        }**

        if (this.userSettings.IsDataMaskingEnabled__c == false || String.isBlank(dataInput) == true) {
            return dataInput;
        }

        for (LogEntryDataMaskRule__mdt dataMaskRule : CACHED_DATA_MASK_RULES.values()) {
            if (dataMaskRule.IsEnabled__c == true) {
                dataInput = dataInput.replaceAll(dataMaskRule.SensitiveDataRegEx__c, dataMaskRule.ReplacementRegEx__c);
            }
        }
       
        **if (String.isNotBlank(dataInput)) {
            // After data masking, trim again the max size
            dataInput = dataInput.left(131072);
        }**
       
        return dataInput;
    }

This limits the text size at before and after data masking.

This solves the problem with very long Logger texts.

**Comment by TrangOul** [comment]
Please do not hardcode the max length. The max for long text may be changed in the future, and it can also differ between fields. Ideally you should check that in the Schema, but you'd need to know which field will these data be saved into.

**Comment by dvdjg** [comment]
The idea I mention is intended to show a general solution, limiting the maximum size for a long text field in Salesforce, not for any other type of sizes. 
    if (String.isNotBlank(dataInput)) {
        dataInput = dataInput.left(131072);
    }
I would prefer if the author took the idea and implemented a more generic solution, but even then this code would be correct because it limits the maximum possible size case in Salesforce.

**Comment by jongpie** [comment]
@BastienRaoul I think I've fixed this issue [in the new release `v4.12.6`](https://github.com/jongpie/NebulaLogger/releases/tag/v4.12.6). I've updated the `LogEntryEventBuilder` class to always first truncate String parameters before trying to apply data mask rules, which should avoid the "regex too complicated" error.

Whenever you have a chance to test it out, let me know if you run into any other errors or issues!
